src/prop/cnf_stream.cpp

```diff
@@ -485,47 +468,93 @@ SatLiteral CnfStream::handleIte(TNode iteNode)
   assertClause(iteNode, iteLit, ~thenLit, ~elseLit);
   assertClause(iteNode, iteLit, ~condLit, ~thenLit);
   assertClause(iteNode, iteLit, condLit, ~elseLit);
-
-  return iteLit;
 }
 
 SatLiteral CnfStream::toCNF(TNode node, bool negated)
 {
   Trace("cnf") << "toCNF(" << node
                << ", negated = " << (negated ? "true" : "false") << ")\n";
+
+  TNode cur;
   SatLiteral nodeLit;
-  Node negatedNode = node.notNode();
-
-  // If the non-negated node has already been translated, get the translation
-  if(hasLiteral(node)) {
-    Trace("cnf") << "toCNF(): already translated\n";
-    nodeLit = getLiteral(node);
-    // Return the (maybe negated) literal
-    return !negated ? nodeLit : ~nodeLit;
-  }
-  // Handle each Boolean operator case
-  switch (node.getKind())
+  std::vector<TNode> visit;
+  std::unordered_map<TNode, bool> cache;
+
+  visit.push_back(node);
+  while (!visit.empty())
   {
-    case kind::NOT: nodeLit = ~toCNF(node[0]); break;
-    case kind::XOR: nodeLit = handleXor(node); break;
-    case kind::ITE: nodeLit = handleIte(node); break;
-    case kind::IMPLIES: nodeLit = handleImplies(node); break;
-    case kind::OR: nodeLit = handleOr(node); break;
-    case kind::AND: nodeLit = handleAnd(node); break;
-    case kind::EQUAL:
-      nodeLit =
-          node[0].getType().isBoolean() ? handleIff(node) : convertAtom(node);
-      break;
-    default:
+    cur = visit.back();
+    Assert(cur.getType().isBoolean());
+
+    if (hasLiteral(cur))
     {
-      nodeLit = convertAtom(node);
+      visit.pop_back();
+      continue;
     }
-    break;
+
+    auto it = cache.find(cur);
+    if (it == cache.end())
+    {
+      cache.emplace(cur, false);
+      Kind k = cur.getKind();
+      // Only traverse Boolean nodes
+      if (k == kind::NOT || k == kind::XOR || k == kind::ITE
+          || k == kind::IMPLIES || k == kind::OR || k == kind::AND
+          || (k == kind::EQUAL && cur[0].getType().isBoolean()))
+      {
+        // Preserve the order of the recursive version
+        for (size_t i = 0, size = cur.getNumChildren(); i < size; ++i)
+        {
+          visit.push_back(cur[size - 1 - i]);
+        }
```

Comment:
```suggestion
        // Preserve the order of the recursive version
        visit.insert(visit.end(), std::make_reverse_iterator(cur.end()), std::make_reverse_iterator(cur.begin()));      
```
(untested)

Reply:
Unfortunately doesn't work with the current `NodeValue::iterator`

```
In file included from /usr/include/c++/11.1.0/bits/stl_algobase.h:67,                                                                                          
                 from /usr/include/c++/11.1.0/bits/char_traits.h:39,                                                                                           
                 from /usr/include/c++/11.1.0/ios:40,                         
                 from /usr/include/c++/11.1.0/ostream:38,                                                                                                      
                 from /home/ma/git/cvc5-fork/src/./base/check.h:37,                                                                                            
                 from /home/ma/git/cvc5-fork/src/./context/cdhashset.h:21,                                                                                     
                 from /home/ma/git/cvc5-fork/src/./prop/cnf_stream.h:29,                                                                                       
                 from /home/ma/git/cvc5-fork/src/prop/cnf_stream.cpp:16:                                                                                       
/usr/include/c++/11.1.0/bits/stl_iterator.h: In instantiation of ‘constexpr std::reverse_iterator<_Iterator>::reference std::reverse_iterator<_Iterator>::operator*() const [with _Iterator = cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> >; std::reverse_iterator<_Iterator>::reference = cvc5::NodeTemplate<false>&]’:                                                                                                                                                        
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:92:45:   required from ‘static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; _ForwardIterator = cvc5::NodeTemplate<false>*; bool _TrivialValueTypes = false]’                                                                                  
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:151:15:   required from ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; _ForwardIterator = cvc5::NodeTemplate<false>*]’                                                                                                                                                              
/usr/include/c++/11.1.0/bits/stl_uninitialized.h:333:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; _ForwardIterator = cvc5::NodeTemplate<false>*; _Tp = cvc5::NodeTemplate<false>]’                                                                                                   
/usr/include/c++/11.1.0/bits/vector.tcc:751:34:   required from ‘void std::vector<_Tp, _Alloc>::_M_range_insert(std::vector<_Tp, _Alloc>::iterator, _ForwardIterator, _ForwardIterator, std::forward_iterator_tag) [with _ForwardIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; _Tp = cvc5::NodeTemplate<false>; _Alloc = std::allocator<cvc5::NodeTemplate<false> >; std::vector<_Tp, _Alloc>::iterator = std::vector<cvc5::NodeTemplate<false> >::iterator]’
/usr/include/c++/11.1.0/bits/stl_vector.h:1665:19:   required from ‘void std::vector<_Tp, _Alloc>::_M_insert_dispatch(std::vector<_Tp, _Alloc>::iterator, _InputIterator, _InputIterator, std::__false_type) [with _InputIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; _Tp = cvc5::NodeTemplate<false>; _Alloc = std::allocator<cvc5::NodeTemplate<false> >; std::vector<_Tp, _Alloc>::iterator = std::vector<cvc5::NodeTemplate<false> >::iterator]’
/usr/include/c++/11.1.0/bits/stl_vector.h:1383:22:   required from ‘std::vector<_Tp, _Alloc>::iterator std::vector<_Tp, _Alloc>::insert(std::vector<_Tp, _Alloc>::const_iterator, _InputIterator, _InputIterator) [with _InputIterator = std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >; <
template-parameter-2-2> = void; _Tp = cvc5::NodeTemplate<false>; _Alloc = std::allocator<cvc5::NodeTemplate<false> >; std::vector<_Tp, _Alloc>::iterator = std::vector<cvc5::NodeTemplate<false> >::iterator; std::vector<_Tp, _Alloc>::const_iterator = std::vector<cvc5::NodeTemplate<false> >::const_iterator]’
/home/ma/git/cvc5-fork/src/prop/cnf_stream.cpp:506:21:   required from here                                                                                    
/usr/include/c++/11.1.0/bits/stl_iterator.h:241:19: error: cannot bind non-const lvalue reference of type ‘std::reverse_iterator<cvc5::expr::NodeValue::iterator<cvc5::NodeTemplate<false> > >::reference’ {aka ‘cvc5::NodeTemplate<false>&’} to an rvalue of type ‘cvc5::NodeTemplate<false>’
  241 |         return *--__tmp;                                                                                                                               
      |                   ^~~~~                                                                                                                                
cc1plus: note: unrecognized command-line option ‘-Wno-unused-private-field’ may have been intended to silence earlier diagnostics                              
make[2]: *** [src/CMakeFiles/cvc5.dir/build.make:1294: src/CMakeFiles/cvc5.dir/prop/cnf_stream.cpp.o] Error 1
```
I don't think that I want to change that in this PR.